code/controllers/subsystem/tts.dm

```diff
@@ -0,0 +1,227 @@
+#define TARGET_INDEX 1
+#define IDENTIFIER_INDEX 2
+#define START_TIME_INDEX 3
+#define REQUEST_INDEX 4
+#define MESSAGE_INDEX 5
+#define EXTRA_TARGETS_INDEX 6
+#define LANGUAGE_INDEX 7
+#define LOCAL_INDEX 8
+#define MESSAGE_RANGE_INDEX 9
+
+SUBSYSTEM_DEF(tts)
+	name = "Text To Speech"
+	wait = 0.05 SECONDS
+	priority = FIRE_PRIORITY_TTS
+	runlevels = RUNLEVEL_LOBBY | RUNLEVEL_SETUP | RUNLEVEL_GAME | RUNLEVEL_POSTGAME
+
+	/// Queued HTTP requests that have yet to be sent
+	var/list/queued_tts_messages = list()
+
+	/// HTTP requests currently in progress but not being processed yet
+	var/list/in_process_tts_messages = list()
+
+	/// HTTP requests that are being processed to see if they've been finished
+	var/list/current_processing_tts_messages = list()
+
+	/// A list of available speakers
+	var/list/available_speakers = list()
+
+	var/list/cached_voices = list()
+
+	/// Whether TTS is enabled or not
+	var/tts_enabled = FALSE
+
+	var/message_timeout = 7 SECONDS
+
+	/// Messages can be timed out earlier if the algorithm thinks that
+	/// it's going to take too long for their message to be processed.
+	/// This'll determine the minimum extent of how late it is allowed to begin timing messages out
+	var/message_timeout_early_minimum = 5 SECONDS
+
+	var/max_concurrent_requests = 4
+
+	/// Used to calculate the average time it takes for a tts message to be received from the http server
+	/// For tts messages which time out, it won't keep tracking the tts message and will just assume that the message took
+	/// 7 seconds (or whatever the value of message_timeout is) to receive back a response.
+	var/average_tts_messages_time = 0
+
+/proc/cmp_word_length_asc(list/a, list/b)
+	return length(a[MESSAGE_INDEX]) - length(b[MESSAGE_INDEX])
+
+/datum/controller/subsystem/tts/vv_edit_var(var_name, var_value)
+	// tts being enabled depends on whether it actually exists
+	if(NAMEOF(src, tts_enabled) == var_name)
+		return FALSE
+	return ..()
+
+/datum/controller/subsystem/tts/stat_entry(msg)
+	msg = "Active:[length(in_process_tts_messages)]|Standby:[length(queued_tts_messages)]|Avg:[average_tts_messages_time]"
+	return ..()
+
+/datum/controller/subsystem/tts/Initialize()
+	if(!CONFIG_GET(string/tts_http_url))
+		return SS_INIT_NO_NEED
+
+	var/datum/http_request/request = new()
+	var/list/headers = list()
+	headers["Authorization"] = CONFIG_GET(string/tts_http_token)
+	request.prepare(RUSTG_HTTP_METHOD_GET, "[CONFIG_GET(string/tts_http_url)]/tts-voices", "", headers)
+	request.begin_async()
+	UNTIL(request.is_complete())
```

Comment:
what is the point of this being async if we're just gonna wait for it ourselves?

Reply:
also this will be usable on a local right?

Reply:
the fact that blocking will block the game's execution. The difference here is that we're sleeping and not pausing the entire server's processing.

Reply:
aight (and the local thing is true?)

Reply:
yes, you can use it on a local as long as you have docker installed. There's instructions on how to run the tts server in the tools/tts folder.
I'm able to test and run the tts code I've written here.